Add missing pmp valorisation to MO production. (Issue #16072)

fappels opened the pull request:
Add pmp valorisation to MO production. (Issue #16072)
Add visualization of product pmp and unit cost update for produced product.
Only when user has pricing knowledge (if he has rights to create bom's)
Add production unitcost calculation for free consumption line
Hide add free consumptionline if production started.
Other bugs: 
mo line is line so no fetchLines
create MO, Column 'tms' cannot be null

fappels committed:
mo line is line so no fetchLines

fappels committed:
Add missing pmp valorisation to MO production. (Issue #16072)

fappels committed:
create MO, Column 'tms' cannot be null

eldy commented on htdocs/mrp/mo_production.php:
```diff
@@ -149,7 +151,7 @@
 		$moline->qty = GETPOST('qtytoadd', 'int'); ;
 		$moline->fk_product = GETPOST('productidtoadd', 'int');
-		$moline->role = 'toconsume';
+		$moline->role = 'toconsumef'; // free consume line
```
Why did you changed key "toconsume"  into  "toconsumef"
A lot of other code is using "toconsume" to calculate virtual stock so key is important.

fappels replied:
I changed to see when line comes from bom or from 'addconsumeline' to add also lines cost from 'addconsumeline' to production unitcost.
First PR was on 12 and wanted to avoid database change, but now for develop branch maybe add new field in llx_mrp_production for this or maybe you have better idea to solve this?
Also not really understand this feature 'addconsumeline', why change bom after manufacturing order started?

eldy replied:
Yes adding new fields is better. For example, you can add the 2 fields
origin_id (id of bom or empty) + origin_type ('bom', 'free', ...)

fappels replied:
Ok will add new fields. Will also check 'addconsumeline' again, I removed add line when production started, after your explication better only remove when production closed.

fappels replied:
Add new fields and allow add free line after production started.
New we have origin id, I replaced show consumed pmp by consumed unitcost.

fappels committed:
Update database with new fields

fappels committed:
Use new origin field to mark free lines.
Allow add free line when production started

fappels committed:
Show also unit cost iso pmp value for consumption

fappels committed:
Merge remote-tracking branch 'Dolibarr/develop' into develop_mo_pmp

fappels committed:
add newline at end of file

fappels committed:
Show consumption unitcost as price.

fappels commented:
Put in draft, still some cost calculation to verify.

eldy committed:
Update mo_production.php

fappels committed:
Merge remote-tracking branch 'Dolibarr/develop' into develop_mo_pmp

eldy commented:
@fappels Can i merge it or is it still in draft ?

fappels commented:
Still in draft, thinking on moving cost calculation from view to actions. 
For example make service is shorter or longer the defined in bom. Cost for margin calculation is different from bom cost. Other solution is only allow diff from bom qty´s by using additional free line with positive or negative qty. @eldy

eldy commented:
We can progress is small step.
A first step to just add the input field, prefilled with cost of bom value is a good first step.

fappels committed:
Merge remote-tracking branch 'Dolibarr/develop' into develop_mo_pmp

fappels committed:
Fix displayed unit cost of free consumption line

fappels commented:
@eldy I fixed some issue on display cost of free line and put out of draft. Ok for best step by step. 
You can modify prefilled bom cost input field, or add add free line.
Example:
![80f2524c-9899-4fc7-affd-0185521e6211](https://user-images.githubusercontent.com/4573808/109019280-25321a80-76b9-11eb-9c45-5480425e9099.png)
Added free line with negative qty for make product, product was made in half time then make time from bom.

fappels committed:
Also register movement for 0 produced to update cost.
Use correct movement label for negative consumed.

fappels commented:
@eldy new commit for correct free line cost update and movement:
Also register movement for 0 produced to update cost.
Use correct movement label for negative consumed.

eldy commented on htdocs/mrp/class/mo.class.php:
```diff
@@ -655,6 +655,8 @@ public function updateProduction(User $user, $notrigger = true)
 
 							$moline->fk_mo = $this->id;
+							$moline->origin_id = $line->id;
+							$moline->origin_type = 'bom';
```
Is it id of BOM or id of line in a BOM ?
If id of line in a BOM, the origin_type = 'bom_line' is better.

fappels replied:
Should I do $moline->origin_type = $line->element; ?

eldy commented on htdocs/mrp/mo_production.php:
```diff
@@ -245,36 +252,35 @@
+							setEventMessages($stockmove->error, $stockmove->errors, 'errors');
+						}
+					}
+					if ($qtytoprocess != 0) {
```
Why removing this test ?

fappels replied:
I moved check below stock movement to be able to do 0 stock movement of produced product and update produced pmp with some free consume line.

fappels committed:
MoLine origin is bomline not bom

homer8173 commented:
@eldy 
Excuse me to reopen this subject but the pmp is still wrong in development branche
We can't send a stock movement with a pmp at 0, it destroy everything in calculation of cost 
because as the pmp may evolve in time, the production cost must use the pmp stored in stock movement
